crypto/polyval-generic.c

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * POLYVAL: hash function for GCM-SIV.
+ * Code based on crypto/ghash-generic.c
+ *
+ * POLYVAL is a keyed hash function similar to GHASH. POLYVAL uses a
+ * different modulus for finite field multiplication which makes hardware
+ * accelerated implementations on little-endian machines faster.
+ *
+ * Like GHASH, POLYVAL is not a cryptographic hash function and should
+ * not be used outside of crypto modes explicitly designed to use POLYVAL.
+ *
+ * Copyright (c) 2007 Nokia Siemens Networks - Mikko Herranen <mh1@iki.fi>
```

Comment:
Copyright lines should be closer to the top of the file.  See crypto/adiantum.c for a good way to do it.  Order is SPDX license identifier, then a block comment containing a 1-line summary of the file and the copyright line(s), and then a separate block comment with a long description of the file.